// @jsonforms/angular-material/lib/cjs/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonFormsAngularMaterialModule = void 0;
/*
  The MIT License

  Copyright (c) 2017-2019 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var common_1 = require("@angular/common");
var flex_layout_1 = require("@angular/flex-layout");
var forms_1 = require("@angular/forms");
var autocomplete_1 = require("@angular/material/autocomplete");
var button_1 = require("@angular/material/button");
var badge_1 = require("@angular/material/badge");
var card_1 = require("@angular/material/card");
var checkbox_1 = require("@angular/material/checkbox");
var datepicker_1 = require("@angular/material/datepicker");
var form_field_1 = require("@angular/material/form-field");
var tooltip_1 = require("@angular/material/tooltip");
var icon_1 = require("@angular/material/icon");
var input_1 = require("@angular/material/input");
var list_1 = require("@angular/material/list");
var core_1 = require("@angular/material/core");
var select_1 = require("@angular/material/select");
var sidenav_1 = require("@angular/material/sidenav");
var slider_1 = require("@angular/material/slider");
var slide_toggle_1 = require("@angular/material/slide-toggle");
var table_1 = require("@angular/material/table");
var tabs_1 = require("@angular/material/tabs");
var toolbar_1 = require("@angular/material/toolbar");
var core_2 = require("@angular/core");
var angular_1 = require("@jsonforms/angular");
var autocomplete_renderer_1 = require("./controls/autocomplete.renderer");
var boolean_renderer_1 = require("./controls/boolean.renderer");
var date_renderer_1 = require("./controls/date.renderer");
var number_renderer_1 = require("./controls/number.renderer");
var range_renderer_1 = require("./controls/range.renderer");
var textarea_renderer_1 = require("./controls/textarea.renderer");
var text_renderer_1 = require("./controls/text.renderer");
var toggle_renderer_1 = require("./controls/toggle.renderer");
var label_renderer_1 = require("./other/label.renderer");
var detail_1 = require("./other/master-detail/detail");
var master_1 = require("./other/master-detail/master");
var object_renderer_1 = require("./other/object.renderer");
var table_renderer_1 = require("./other/table.renderer");
var categorization_layout_renderer_1 = require("./layouts/categorization-layout.renderer");
var group_layout_renderer_1 = require("./layouts/group-layout.renderer");
var horizontal_layout_renderer_1 = require("./layouts/horizontal-layout.renderer");
var vertical_layout_renderer_1 = require("./layouts/vertical-layout.renderer");
var array_layout_renderer_1 = require("./layouts/array-layout.renderer");
var JsonFormsAngularMaterialModule = /** @class */ (function () {
    function JsonFormsAngularMaterialModule() {
    }
    JsonFormsAngularMaterialModule.decorators = [
        { type: core_2.NgModule, args: [{
                    imports: [
                        common_1.CommonModule,
                        angular_1.JsonFormsModule,
                        form_field_1.MatFormFieldModule,
                        checkbox_1.MatCheckboxModule,
                        input_1.MatInputModule,
                        slider_1.MatSliderModule,
                        slide_toggle_1.MatSlideToggleModule,
                        core_1.MatNativeDateModule,
                        datepicker_1.MatDatepickerModule,
                        tabs_1.MatTabsModule,
                        sidenav_1.MatSidenavModule,
                        list_1.MatListModule,
                        forms_1.ReactiveFormsModule,
                        flex_layout_1.FlexLayoutModule,
                        card_1.MatCardModule,
                        select_1.MatSelectModule,
                        button_1.MatButtonModule,
                        icon_1.MatIconModule,
                        autocomplete_1.MatAutocompleteModule,
                        table_1.MatTableModule,
                        toolbar_1.MatToolbarModule,
                        tooltip_1.MatTooltipModule,
                        badge_1.MatBadgeModule
                    ],
                    declarations: [
                        boolean_renderer_1.BooleanControlRenderer,
                        textarea_renderer_1.TextAreaRenderer,
                        text_renderer_1.TextControlRenderer,
                        number_renderer_1.NumberControlRenderer,
                        range_renderer_1.RangeControlRenderer,
                        date_renderer_1.DateControlRenderer,
                        toggle_renderer_1.ToggleControlRenderer,
                        vertical_layout_renderer_1.VerticalLayoutRenderer,
                        horizontal_layout_renderer_1.HorizontalLayoutRenderer,
                        categorization_layout_renderer_1.CategorizationTabLayoutRenderer,
                        group_layout_renderer_1.GroupLayoutRenderer,
                        label_renderer_1.LabelRenderer,
                        master_1.MasterListComponent,
                        detail_1.JsonFormsDetailComponent,
                        object_renderer_1.ObjectControlRenderer,
                        autocomplete_renderer_1.AutocompleteControlRenderer,
                        table_renderer_1.TableRenderer,
                        array_layout_renderer_1.ArrayLayoutRenderer
                    ],
                    entryComponents: [
                        boolean_renderer_1.BooleanControlRenderer,
                        textarea_renderer_1.TextAreaRenderer,
                        text_renderer_1.TextControlRenderer,
                        number_renderer_1.NumberControlRenderer,
                        range_renderer_1.RangeControlRenderer,
                        date_renderer_1.DateControlRenderer,
                        toggle_renderer_1.ToggleControlRenderer,
                        vertical_layout_renderer_1.VerticalLayoutRenderer,
                        horizontal_layout_renderer_1.HorizontalLayoutRenderer,
                        group_layout_renderer_1.GroupLayoutRenderer,
                        label_renderer_1.LabelRenderer,
                        categorization_layout_renderer_1.CategorizationTabLayoutRenderer,
                        master_1.MasterListComponent,
                        detail_1.JsonFormsDetailComponent,
                        object_renderer_1.ObjectControlRenderer,
                        autocomplete_renderer_1.AutocompleteControlRenderer,
                        table_renderer_1.TableRenderer,
                        array_layout_renderer_1.ArrayLayoutRenderer
                    ],
                    exports: [
                        common_1.CommonModule,
                        angular_1.JsonFormsModule,
                        form_field_1.MatFormFieldModule,
                        checkbox_1.MatCheckboxModule,
                        input_1.MatInputModule,
                        slider_1.MatSliderModule,
                        slide_toggle_1.MatSlideToggleModule,
                        core_1.MatNativeDateModule,
                        datepicker_1.MatDatepickerModule,
                        tabs_1.MatTabsModule,
                        sidenav_1.MatSidenavModule,
                        list_1.MatListModule,
                        forms_1.ReactiveFormsModule,
                        flex_layout_1.FlexLayoutModule,
                        card_1.MatCardModule,
                        select_1.MatSelectModule,
                        button_1.MatButtonModule,
                        icon_1.MatIconModule,
                        autocomplete_1.MatAutocompleteModule
                    ],
                    schemas: [core_2.CUSTOM_ELEMENTS_SCHEMA],
                    providers: []
                },] }
    ];
    return JsonFormsAngularMaterialModule;
}());
exports.JsonFormsAngularMaterialModule = JsonFormsAngularMaterialModule;
