// @jsonforms/angular/lib/esm/jsonforms-root.component.js
/*
  The MIT License

  Copyright (c) 2017-2020 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Actions } from '@jsonforms/core';
import Ajv from 'ajv';
import { JsonFormsAngularService, USE_STATE_VALUE } from './jsonforms.service';
var JsonForms = /** @class */ (function () {
    function JsonForms(jsonformsService) {
        this.jsonformsService = jsonformsService;
        this.dataChange = new EventEmitter();
        this.errors = new EventEmitter();
        this.initialized = false;
    }
    JsonForms.prototype.ngOnInit = function () {
        var _this = this;
        this.jsonformsService.init({
            core: {
                data: this.data,
                uischema: this.uischema,
                schema: this.schema,
                ajv: this.ajv,
                validationMode: this.validationMode,
                additionalErrors: this.additionalErrors
            },
            uischemas: this.uischemas,
            i18n: this.i18n,
            renderers: this.renderers,
            config: this.config,
            readonly: this.readonly
        });
        this.jsonformsService.$state.subscribe(function (state) {
            var _a, _b, _c, _d;
            var data = (_b = (_a = state === null || state === void 0 ? void 0 : state.jsonforms) === null || _a === void 0 ? void 0 : _a.core) === null || _b === void 0 ? void 0 : _b.data;
            var errors = (_d = (_c = state === null || state === void 0 ? void 0 : state.jsonforms) === null || _c === void 0 ? void 0 : _c.core) === null || _d === void 0 ? void 0 : _d.errors;
            if (_this.previousData !== data) {
                _this.previousData = data;
                _this.dataChange.emit(data);
            }
            if (_this.previousErrors !== errors) {
                _this.previousErrors = errors;
                _this.errors.emit(errors);
            }
        });
        this.oldI18N = this.i18n;
        this.initialized = true;
    };
    JsonForms.prototype.ngDoCheck = function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
        // we can't use ngOnChanges as then nested i18n changes will not be detected
        // the update will result in a no-op when the parameters did not change
        if (((_a = this.oldI18N) === null || _a === void 0 ? void 0 : _a.locale) !== ((_b = this.i18n) === null || _b === void 0 ? void 0 : _b.locale) ||
            ((_c = this.oldI18N) === null || _c === void 0 ? void 0 : _c.translate) !== ((_d = this.i18n) === null || _d === void 0 ? void 0 : _d.translate) ||
            ((_e = this.oldI18N) === null || _e === void 0 ? void 0 : _e.translateError) !== ((_f = this.i18n) === null || _f === void 0 ? void 0 : _f.translateError)) {
            this.jsonformsService.updateI18n(Actions.updateI18n(((_g = this.oldI18N) === null || _g === void 0 ? void 0 : _g.locale) === ((_h = this.i18n) === null || _h === void 0 ? void 0 : _h.locale)
                ? this.jsonformsService.getState().jsonforms.i18n.locale
                : (_j = this.i18n) === null || _j === void 0 ? void 0 : _j.locale, ((_k = this.oldI18N) === null || _k === void 0 ? void 0 : _k.translate) === ((_l = this.i18n) === null || _l === void 0 ? void 0 : _l.translate)
                ? this.jsonformsService.getState().jsonforms.i18n.translate
                : (_m = this.i18n) === null || _m === void 0 ? void 0 : _m.translate, ((_o = this.oldI18N) === null || _o === void 0 ? void 0 : _o.translateError) === ((_p = this.i18n) === null || _p === void 0 ? void 0 : _p.translateError)
                ? this.jsonformsService.getState().jsonforms.i18n.translateError
                : (_q = this.i18n) === null || _q === void 0 ? void 0 : _q.translateError));
            this.oldI18N = this.i18n;
        }
    };
    // tslint:disable-next-line: cyclomatic-complexity
    JsonForms.prototype.ngOnChanges = function (changes) {
        var _a, _b, _c;
        if (!this.initialized) {
            return;
        }
        var newData = changes.data;
        var newSchema = changes.schema;
        var newUiSchema = changes.uischema;
        var newRenderers = changes.renderers;
        var newUischemas = changes.uischemas;
        var newI18n = changes.i18n;
        var newReadonly = changes.readonly;
        var newValidationMode = changes.validationMode;
        var newAjv = changes.ajv;
        var newConfig = changes.config;
        var newAdditionalErrors = changes.additionalErrors;
        if (newData || newSchema || newUiSchema || newValidationMode || newAjv || newAdditionalErrors) {
            this.jsonformsService.updateCoreState(newData ? newData.currentValue : USE_STATE_VALUE, newSchema ? newSchema.currentValue : USE_STATE_VALUE, newUiSchema ? newUiSchema.currentValue : USE_STATE_VALUE, newAjv ? newAjv.currentValue : USE_STATE_VALUE, newValidationMode ? newValidationMode.currentValue : USE_STATE_VALUE, newAdditionalErrors ? newAdditionalErrors.currentValue : USE_STATE_VALUE);
        }
        if (newRenderers && !newRenderers.isFirstChange()) {
            this.jsonformsService.setRenderers(newRenderers.currentValue);
        }
        if (newUischemas && !newUischemas.isFirstChange()) {
            this.jsonformsService.setUiSchemas(newUischemas.currentValue);
        }
        if (newI18n && !newI18n.isFirstChange()) {
            this.jsonformsService.updateI18n(Actions.updateI18n((_a = newI18n.currentValue) === null || _a === void 0 ? void 0 : _a.locale, (_b = newI18n.currentValue) === null || _b === void 0 ? void 0 : _b.translate, (_c = newI18n.currentValue) === null || _c === void 0 ? void 0 : _c.translateError));
        }
        if (newReadonly && !newReadonly.isFirstChange()) {
            this.jsonformsService.setReadonly(newReadonly.currentValue);
        }
        if (newConfig && !newConfig.isFirstChange()) {
            this.jsonformsService.updateConfig(Actions.setConfig(newConfig.currentValue));
        }
    };
    JsonForms.decorators = [
        { type: Component, args: [{
                    selector: 'jsonforms',
                    template: '<jsonforms-outlet></jsonforms-outlet>',
                    providers: [JsonFormsAngularService]
                },] }
    ];
    JsonForms.ctorParameters = function () { return [
        { type: JsonFormsAngularService }
    ]; };
    JsonForms.propDecorators = {
        uischema: [{ type: Input }],
        schema: [{ type: Input }],
        data: [{ type: Input }],
        renderers: [{ type: Input }],
        uischemas: [{ type: Input }],
        dataChange: [{ type: Output }],
        readonly: [{ type: Input }],
        validationMode: [{ type: Input }],
        ajv: [{ type: Input }],
        config: [{ type: Input }],
        i18n: [{ type: Input }],
        additionalErrors: [{ type: Input }],
        errors: [{ type: Output }]
    };
    return JsonForms;
}());
export { JsonForms };
