// @jsonforms/angular/lib/esm/jsonforms.component.js
import { __extends } from "tslib";
/*
  The MIT License

  Copyright (c) 2017-2019 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
import maxBy from 'lodash/maxBy';
import { ComponentFactoryResolver, Directive, Input, ViewContainerRef } from '@angular/core';
import { createId, isControl, getConfig, mapStateToJsonFormsRendererProps } from '@jsonforms/core';
import { UnknownRenderer } from './unknown.component';
import { JsonFormsBaseRenderer } from './base.renderer';
import { JsonFormsControl } from './control';
import { JsonFormsAngularService } from './jsonforms.service';
import { get, isEqual } from 'lodash';
var areEqual = function (prevProps, nextProps) {
    return get(prevProps, 'renderers.length') === get(nextProps, 'renderers.length')
        && get(prevProps, 'cells.length') === get(nextProps, 'cells.length')
        && get(prevProps, 'uischemas.length') === get(nextProps, 'uischemas.length')
        && get(prevProps, 'schema') === get(nextProps, 'schema')
        && isEqual(get(prevProps, 'uischema'), get(nextProps, 'uischema'))
        && get(prevProps, 'path') === get(nextProps, 'path');
};
var ɵ0 = areEqual;
var JsonFormsOutlet = /** @class */ (function (_super) {
    __extends(JsonFormsOutlet, _super);
    function JsonFormsOutlet(viewContainerRef, componentFactoryResolver, jsonformsService) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.jsonformsService = jsonformsService;
        return _this;
    }
    Object.defineProperty(JsonFormsOutlet.prototype, "renderProps", {
        set: function (renderProps) {
            this.path = renderProps.path;
            this.schema = renderProps.schema;
            this.uischema = renderProps.uischema;
            this.update(this.jsonformsService.getState());
        },
        enumerable: false,
        configurable: true
    });
    JsonFormsOutlet.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.jsonformsService.$state.subscribe({
            next: function (state) { return _this.update(state); }
        });
    };
    JsonFormsOutlet.prototype.update = function (state) {
        var props = mapStateToJsonFormsRendererProps(state, {
            schema: this.schema,
            uischema: this.uischema,
            path: this.path
        });
        if (areEqual(this.previousProps, props)) {
            return;
        }
        else {
            this.previousProps = props;
        }
        var renderers = props.renderers;
        var schema = this.schema || props.schema;
        var uischema = this.uischema || props.uischema;
        var testerContext = {
            rootSchema: props.rootSchema,
            config: getConfig(state)
        };
        var renderer = maxBy(renderers, function (r) { return r.tester(uischema, schema, testerContext); });
        var bestComponent = UnknownRenderer;
        if (renderer !== undefined && renderer.tester(uischema, schema, testerContext) !== -1) {
            bestComponent = renderer.renderer;
        }
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(bestComponent);
        this.viewContainerRef.clear();
        var currentComponentRef = this.viewContainerRef.createComponent(componentFactory);
        if (currentComponentRef.instance instanceof JsonFormsBaseRenderer) {
            var instance = currentComponentRef.instance;
            instance.uischema = uischema;
            instance.schema = schema;
            instance.path = this.path;
            if (instance instanceof JsonFormsControl) {
                var controlInstance = instance;
                if (controlInstance.id === undefined) {
                    var id = isControl(props.uischema)
                        ? createId(props.uischema.scope)
                        : undefined;
                    instance.id = id;
                }
            }
        }
    };
    JsonFormsOutlet.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    JsonFormsOutlet.decorators = [
        { type: Directive, args: [{
                    selector: 'jsonforms-outlet'
                },] }
    ];
    JsonFormsOutlet.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ComponentFactoryResolver },
        { type: JsonFormsAngularService }
    ]; };
    JsonFormsOutlet.propDecorators = {
        renderProps: [{ type: Input }]
    };
    return JsonFormsOutlet;
}(JsonFormsBaseRenderer));
export { JsonFormsOutlet };
export { ɵ0 };
