// @jsonforms/angular-material/lib/cjs/controls/autocomplete.renderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.enumControlTester = exports.AutocompleteControlRenderer = void 0;
var tslib_1 = require("tslib");
/*
  The MIT License
  
  Copyright (c) 2017-2019 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var core_1 = require("@angular/core");
var angular_1 = require("@jsonforms/angular");
var core_2 = require("@jsonforms/core");
var operators_1 = require("rxjs/operators");
var operators_2 = require("rxjs/operators");
/**
 * To use this component you will need to add your own tester:
 * <pre><code>
 * ...
 * export const AutocompleteControlRendererTester: RankedTester = rankWith(2, isEnumControl);
 * ...
 * </code></pre>
 * Add the tester and renderer to JSONForms registry:
 * <pre><code>
 * ...
 * { tester: AutocompleteControlRendererTester, renderer: AutocompleteControlRenderer },
 * ...
 * </code></pre>
 * Furthermore you need to update your module.
 * <pre><code>
 * ...
 * imports: [JsonFormsAngularMaterialModule, MatAutocompleteModule],
 * declarations: [AutocompleteControlRenderer],
 * entryComponents: [AutocompleteControlRenderer]
 * ...
 * </code></pre>
 *
 */
var AutocompleteControlRenderer = /** @class */ (function (_super) {
    tslib_1.__extends(AutocompleteControlRenderer, _super);
    function AutocompleteControlRenderer(jsonformsService) {
        var _this = _super.call(this, jsonformsService) || this;
        _this.getEventValue = function (event) { return event.target.value; };
        return _this;
    }
    AutocompleteControlRenderer.prototype.ngOnInit = function () {
        var _this = this;
        _super.prototype.ngOnInit.call(this);
        this.shouldFilter = false;
        this.filteredOptions = this.form.valueChanges.pipe(operators_2.startWith(''), operators_1.map(function (val) { return _this.filter(val); }));
    };
    AutocompleteControlRenderer.prototype.updateFilter = function (event) {
        // ENTER
        if (event.keyCode === 13) {
            this.shouldFilter = false;
        }
        else {
            this.shouldFilter = true;
        }
    };
    AutocompleteControlRenderer.prototype.onSelect = function (ev) {
        var path = core_2.composeWithUi(this.uischema, this.path);
        this.shouldFilter = false;
        this.jsonFormsService.updateCore(core_2.Actions.update(path, function () { return ev.option.value; }));
        this.triggerValidation();
    };
    AutocompleteControlRenderer.prototype.filter = function (val) {
        var _this = this;
        return (this.options || this.scopedSchema.enum || []).filter(function (option) {
            return !_this.shouldFilter ||
                !val ||
                option.toLowerCase().indexOf(val.toLowerCase()) === 0;
        });
    };
    AutocompleteControlRenderer.prototype.getOwnProps = function () {
        return tslib_1.__assign(tslib_1.__assign({}, _super.prototype.getOwnProps.call(this)), { options: this.options });
    };
    AutocompleteControlRenderer.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'AutocompleteControlRenderer',
                    template: "\n    <mat-form-field fxFlex [fxHide]=\"hidden\">\n      <mat-label>{{ label }}</mat-label>\n      <input\n        matInput\n        type=\"text\"\n        (change)=\"onChange($event)\"\n        [id]=\"id\"\n        [formControl]=\"form\"\n        [matAutocomplete]=\"auto\"\n        (keydown)=\"updateFilter($event)\"\n      />\n      <mat-autocomplete\n        autoActiveFirstOption\n        #auto=\"matAutocomplete\"\n        (optionSelected)=\"onSelect($event)\"\n      >\n        <mat-option\n          *ngFor=\"let option of filteredOptions | async\"\n          [value]=\"option\"\n        >\n          {{ option }}\n        </mat-option>\n      </mat-autocomplete>\n      <mat-hint *ngIf=\"shouldShowUnfocusedDescription()\">{{ description }}</mat-hint>\n      <mat-error>{{ error }}</mat-error>\n    </mat-form-field>\n  ",
                    changeDetection: core_1.ChangeDetectionStrategy.OnPush
                },] }
    ];
    AutocompleteControlRenderer.ctorParameters = function () { return [
        { type: angular_1.JsonFormsAngularService }
    ]; };
    AutocompleteControlRenderer.propDecorators = {
        options: [{ type: core_1.Input }]
    };
    return AutocompleteControlRenderer;
}(angular_1.JsonFormsControl));
exports.AutocompleteControlRenderer = AutocompleteControlRenderer;
exports.enumControlTester = core_2.rankWith(2, core_2.isEnumControl);
